research/query_service/gremlin/scripts/run_gaia.py

```diff
@@ -0,0 +1,247 @@
+parser.add_argument("-P", "--port", help="The port for RPC connection")
+parser.add_argument("-D", "--download", type=str2bool, nargs='?', const=True, default=False,
+                    help="Whether to download raw file")
+# parser.add_argument("-C", "--clean", type=str2bool, nargs='?', const=True, default=False, help="Clean the process")
+
+args = parser.parse_args()
+if args.tmp_dir is not None:
+    tmp_dir = args.tmp_dir
+
+if args.host_file is not None:
+    hosts = hosts_dict_to_list(get_hosts_from_toml(args.host_file))
+    num_machines = args.num_machines
+    if num_machines == 0 or len(hosts) < num_machines:
+        # The default number of machines is the number of hosts in the host file
+        num_machines = len(hosts)
+else:
+    num_machines = 1
+
+####################################################################
+################### End  of preparing arguments ####################
+####################################################################
+
+
+def prepare_files():
+    schema_folder = os.path.join(args.graph_dir, "graph_schema")
+    schema_file = os.path.join(args.graph_dir, "graph_schema", "schema.json")
+
+    subprocess.Popen(["mkdir", "-p", tmp_dir]).wait()
+    subprocess.Popen(["mkdir", "-p", schema_folder]).wait()
+    subprocess.Popen(["cp", args.schema_file, schema_file]).wait()
+
+    if args.host_file is not None:
+        host_list_file = os.path.join(tmp_dir, "hosts")
+        with open(host_list_file, 'w') as f:
+            for host in hosts:
+                f.write("%s:%s\n" % (host[0], host[1]))
+        for i in range(num_machines):
+            subprocess.Popen(["ssh", hosts[i][0], "mkdir", "-p", tmp_dir]).wait()
+            subprocess.Popen(["ssh", hosts[i][0], "mkdir", "-p", schema_folder]).wait()
+            subprocess.Popen(["scp", args.schema_file, "%s:%s" % (hosts[i][0], schema_file)]).wait()
+            subprocess.Popen(["scp", host_list_file, "%s:%s" % (hosts[i][0], tmp_dir)]).wait()
+
+
+def download_from_hdfs():
+    """
+    Download given files from HDFS to local hosts
+    """
+    program = os.path.join(os.getcwd(), "bin", "downloader")
+    pool = []
+    env = "RUST_LOG=Info"
+    if args.host_file is not None:
+        host_list_file = os.path.join(tmp_dir, "hosts")
+        for i in range(num_machines):
+            run_params = "%s %s %s %s %s %d -w %d -n %d -p %d -h %s" % \
+                     (env, program, args.hadoop_home, args.raw_data_dir, tmp_dir,
+                      args.raw_partitions, args.workers, num_machines, i, host_list_file)
+            cmd = ["ssh", hosts[i][0], run_params]
+            print(cmd)
+            pool.append(subprocess.Popen(cmd))
+        sync_process(pool)
+    else:
+        cmd = "%s %s %s %s %s %d -w %d" % (env, program, args.hadoop_home, args.raw_data_dir, tmp_dir,
+                      args.raw_partitions, args.workers)
+        print(cmd)
+        os.system(cmd)
```

Comment:
you can replace `os.system` with:

```py
subprocess.check_call([cmd])
```

Reply:
Hi, without `shell=True`, the script cannot run. Is there a solution
![Screen Shot 2021-12-03 at 14 33 43](https://user-images.githubusercontent.com/7007589/144556576-f9bee7bf-0f04-4b15-a254-bc09fea5ed43.png)
?

Reply:
From [Doc](https://docs.python.org/3/library/subprocess.html#subprocess.check_call):

> check_call: Run command with arguments. Wait for command to complete. If the return code was zero then return, otherwise raise CalledProcessError. The CalledProcessError object will have the return code in the returncode attribute. If check_call() was unable to start the process it will propagate the exception that was raised.

```python
subprocess.check_call(["python3", "/home/admin/xxx.py", "arg1", "arg2"], env=dict(os.environ, PATH=PATH))
```